llama-cpp and ggml: new packages

rbberger opened the pull request:
This is a useful LLM package for quickly running LLMs without many dependencies. Installing it via Spack allows you to then run something like:

```
llama-server -hf ggml-org/gpt-oss-20b-GGUF --jinja -c 0 --host 127.0.0.1 --port 8033
```

A python interface was added in https://github.com/spack/spack-packages/pull/2441

- [x] ~Depends on: https://github.com/spack/spack-packages/pull/2436~ (merged with this one)

Given the high development velocity of llama-cpp, it makes more sense to build with the embedded GGML, which is why both packages share a common base class for adding GGML build options, variants and dependencies.

rbberger committed:
ggml: add new package

rbberger committed:
llama-cpp: new package

aweits commented:
see #2314, can we collab on this?

rbberger committed:
llama-cpp: add LLAMA_CPP_LIB_PATH env var

rbberger committed:
ggml: add rpc variant

Co-authored-by: Andrew Elble <aweits@rit.edu>

rbberger committed:
ggml: use veclibfort as Accelerate placeholder

rbberger committed:
llama-cpp: add patch for rpc tool install

rbberger commented:
@aweits can you try this one out and see if it does what you want? I've used the merge request patch to fix the RPC install.

aweits commented:
I'll look, but i think the patch will have to be carried in the ggml package...

rbberger commented:
@aweits the main issue with the split I see is having a matching ggml release for a llama-cpp release/tag. There's always git versions to make it work with master, but we probably want at least the named versions have a functional counterpart. Should we just add our own for certain sync points?

aweits commented:
> @aweits the main issue with the split I see is having a matching ggml release for a llama-cpp release/tag. There's always git versions to make it work with master, but we probably want at least the named versions have a functional counterpart. Should we just add our own for certain sync points?

I haven't looked at if gitversions have matured enough to use them, but we certainly can put the "versions" we require in there i guess? I'll poke around in a bit...

rbberger committed:
ggml: add latest sync point with llama.cpp

rbberger committed:
ggml: add latest sync point with whisper.cpp

rbberger committed:
llama-cpp: add latest release synced with ggml

aweits commented on repos/spack_repo/builtin/packages/llama_cpp/package.py:
```diff
@@ -0,0 +1,57 @@
+    license("MIT")
+
+    version("master", branch="master")
+    version("7086" tag="b7086", commit="7aaeedc098a77e9323044187101db4f6b69988da")
```
```suggestion
    version("7086", tag="b7086", commit="7aaeedc098a77e9323044187101db4f6b69988da")
```

aweits commented:
ok, this works for me, can we add the rpc variant to llama-cpp as well?

rbberger committed:
Update repos/spack_repo/builtin/packages/llama_cpp/package.py

Co-authored-by: Andrew W Elble <aweits@rit.edu>

rbberger commented:
@aweits you're asking for convenience for doing `llama-cpp+rpc` instead of `llama-cpp ^ggml+rpc`? Since it does seem to work without explicitly setting any more CMake flags, it would only be used to require `ggml+rpc`.

aweits commented:
> @aweits you're asking for convenience for doing `llama-cpp+rpc` instead of `llama-cpp ^ggml+rpc`? Since it does seem to work without explicitly setting any more CMake flags, it would only be used to require `ggml+rpc`.

That makes sense to me - easier for folks to install without having to infer more details around how ggml is/was built.

rbberger committed:
ggml+llama-cpp: share common ggml build logic and prefer embedded ggml

rbberger commented:
@aweits I went ahead and created a common base class for both ggml and llama-cpp. You probably can do the same once this one is merged with whisper.cpp. Given the speed of llama-cpp, I'm defaulting this now to prefer doing the embedded GGML build. This should give us the best of both worlds and a quicker way of running current llama-cpp without having to wait for ggml to get synced. I've dropped the patch for RPC and simply removed the older versions that aren't patched yet.

rbberger committed:
llama-cpp: remove unused import

aweits commented:
> @aweits I went ahead and created a common base class for both ggml and llama-cpp. You probably can do the same once this one is merged with whisper.cpp. Given the speed of llama-cpp, I'm defaulting this now to prefer doing the embedded GGML build. This should give us the best of both worlds and a quicker way of running current llama-cpp without having to wait for ggml to get synced. I've dropped the patch for RPC and simply removed the older versions that aren't patched yet.

I like this a lot, let me do a couple of test builds...

aweits reviewed (dismissed):
LGTM.
```
[aweits@coltrane spack-packages]$ spack find -lv llama-cpp
-- linux-rocky9-skylake / %c,cxx=gcc@13.2.0 ---------------------
52kpqnt llama-cpp@7158+blas+cpu+cuda+curl~ipo+openmp~rocm+rpc+shared~system_ggml build_system=cmake build_type=Release commit=b3b03a7baf387cfeaf56641bd14c06dbd3d2fcf0 cuda_arch:=80,86 generator=make
```

adamjstewart commented on repos/spack_repo/builtin/packages/ggml/package.py:
```diff
@@ -0,0 +1,89 @@
+    version("master", branch="master")
+    version("0.9.4-20251120", commit="781baf2a14d9e0aaee542b2e1bb918bfc4132199")
+    version("0.9.4-20251117", commit="c23776f22d616d8cb635145381cad365bac675e7")
+    version("0.9.4", tag="v0.9.4", commit="72632094336524a9c809e129e8b1c52154543a5a")
```
Should we default to this version because it's a real release?

rbberger replied:
No. It's very outdated.

adamjstewart replied:
I know how fast the ML world works, but 2 months ago is outdated?

rbberger replied:
If you consider their "releases", yes.

rbberger replied:
@adamjstewart I was considering removing all versions except `master`. What you see remaining were my attempts of finding usable versions between a pure llama-cpp and a ggml package. The fake versions are sync points. Since I basically gave up with the sync points and made the internal build the default, probably I can drop the fake ggml versions altogether.

rbberger replied:
I removed the fake versions and only kept "real" releases in ggml.

adamjstewart commented on repos/spack_repo/builtin/packages/llama_cpp/package.py:
```diff
@@ -0,0 +1,61 @@
+    version("7086", tag="b7086", commit="7aaeedc098a77e9323044187101db4f6b69988da")
+
+    variant("curl", default=True, description="use curl for model download")
+    variant("system_ggml", default=False, description="build embedded GGML library")
```
We should always rely on Spack-installed packages if possible. I would personally drop this variant and always build GGML separately. If we keep this variant, I find the current variant name and description to be confusing.

rbberger replied:
Argh, yes forgot to update the description. The variant name matches the cmake option.

rbberger replied:
Better?

adamjstewart commented on repos/spack_repo/builtin/packages/llama_cpp/package.py:
```diff
@@ -0,0 +1,61 @@
+
+    depends_on("ggml@0.9.4-20251117:", when="@7086 +system_ggml")
+
+    for v in ("cpu", "blas", "openmp", "cuda", "rocm", "metal", "rpc"):
+        depends_on(f"ggml+{v}", when=f"+system_ggml +{v}")
+        depends_on(f"ggml~{v}", when=f"+system_ggml ~{v}")
```
Why are these variants needed in `llama-cpp`? We try to avoid variant forwarding whenever possible, otherwise every package would have variants for all of its dependencies combined. Users can always install `llama-cpp ^ggml+cuda`, right?

aweits replied:
That really is confusing for users that might not even be aware of the existence of ggml and want to distribute inference jobs across gpus - I'd prefer to hide the existence of ggml and leave the variants on llama/whisper in that case.

adamjstewart replied:
It just makes this package orders of magnitude more complicated than it needs to be. If y'all want to maintain this complexity, I can't stop you.

rbberger replied:
@adamjstewart I agree. So much that I argue the same as you in https://github.com/spack/spack-packages/issues/2513 (join the fun ;-) ). What you see here is a middle ground after trying what you are proposing.

adamjstewart commented on repos/spack_repo/builtin/packages/llama_cpp/package.py:
```diff
@@ -0,0 +1,61 @@
+from spack.package import *
+
+
+class LlamaCpp(GGMLPackageBase):
```
I'm not sure if this base class is necessary. This package looks rather simple, all of the complexity is in GGML. If we always depend on GGML and keep all the variants there, none of this complexity is needed in llama-cpp.

aweits replied:
Unfortunately, ggml is not always in-sync with what's vendored in whisper-cpp and llama-cpp. Breaking it out ostensibly creates more problems than it solves - thus this solution for now.

adamjstewart replied:
Not in-sync as in a specific commit, or not in-sync as in there are internal modifications not present in GGML? For the former, this is how PyTorch is packaged, we just pin it to commits. For the latter, I wouldn't even allow external dependencies in that case.

rbberger replied:
@adamjstewart I had the same thoughts as you initially. However, llama.cpp and whisper.cpp development is a bit weird. ggml is developed as part of those clients and periodically syncing back to ggml. There are no good current releases in ggml itself. It's worse than a submodule. It's inverted in terms of what's current.

adamjstewart replied:
In that case, we can't guarantee that ggml is ever up-to-date enough to use to build llama-cpp, and there's no way to know which versions are compatible. I would remove the option to build with Spack-installed ggml.

rbberger replied:
The hope would be that as it matures they'll have proper releases. What you see now is an approach to avoid duplication. I do imagine ggml being useful on its own, but I didn't want to maintain the cmake logic in two places.

adamjstewart reviewed (changes requested):
I think this could be greatly simplified to behave more like other Spack packages.

rbberger committed:
llama-cpp: correct system_ggml variant description

rbberger committed:
ggml: remove fake versions

We default to embedded builds in llama-cpp now. So there is no real
value added by fake ggml versions matching sync points. You can still
use @git.<sha>=version to build specific commits.

adamjstewart reviewed (approved):
It isn't necessarily how I would do it, but I don't want to block this PR from being merged.
